ivc/src/poseidon/interpreter.rs

```diff
@@ -0,0 +1,156 @@
+{
+    use Power::*;
+    let pwr = PoseidonColumn::Powers;
+
+    let square = x.clone() * x.clone();
+    let square = env.hcopy(&square, pwr(Square(i)));
+
+    let fourth = square.clone() * square.clone();
+    let fourth = env.hcopy(&fourth, pwr(Fourth(i)));
+
+    let sixth = fourth * square;
+    let sixth = env.hcopy(&sixth, pwr(Sixth(i)));
+
+    let seventh = sixth * x;
+    let seventh = env.hcopy(&seventh, pwr(Seventh(i)));
+    seventh
+}
+
+fn round<F: PrimeField, Env>(
+    env: &mut Env,
+    state: [Env::Variable; 3],
+    absorb: [Env::Variable; 2],
+    check_out: Env::Variable,
+) -> [Env::Variable; 3]
+where
+    F: PrimeField,
+    Env: ColAccessCap<F, PoseidonColumn> + HybridCopyCap<F, PoseidonColumn>,
+{
+    let mds = mds::<F, Env>();
+    let one = Env::constant(F::one());
+    //read round constants
+    let r0 = env.read_column(PoseidonColumn::RoundConstant(0));
+    let r1 = env.read_column(PoseidonColumn::RoundConstant(1));
+    let r2 = env.read_column(PoseidonColumn::RoundConstant(2));
+    //x^7
+    let powers = powers(env, state);
+    //combine the powers with each mds row and add round constant
+    let [mds_0, mds_1, mds_2] = mds;
+    let s0 = combine::<F, Env>(powers.clone(), mds_0) + r0;
+    let s1 = combine::<F, Env>(powers.clone(), mds_1) + r1;
+    let s2 = combine::<F, Env>(powers.clone(), mds_2) + r2;
+
+    //asserting the result if check is enabled
+    let check_selector = env.read_column(PoseidonColumn::Mode(Selector::CheckEnabled));
+    let check_value = check_out;
+    let check = (s0.clone() - check_value) * check_selector;
+    env.assert_zero(check);
+
+    //if init is set, zeros are writtent instead of the round's output
+    let init_selector = one - env.read_column(PoseidonColumn::Mode(Selector::Init));
+    let s0 = s0 * init_selector.clone();
+    let s1 = s1 * init_selector.clone();
+    let s2 = s2 * init_selector;
+
+    //absorbing 2 values if absorb is enabled
+    let absorb_selector = env.read_column(PoseidonColumn::Mode(Selector::Absorb));
+    let [a0, a1] = absorb;
+    let s0 = s0 + a0 * absorb_selector.clone();
+    let s1 = s1 + a1 * absorb_selector;
+
+    let state = [s0, s1, s2];
+    state
+}
+
+///constrains a round
+pub fn constraint_round<F: PrimeField, Env>(env: &mut Env)
+where
+    F: PrimeField,
+    Env: ColAccessCap<F, PoseidonColumn>
+        + HybridCopyCap<F, PoseidonColumn>
+        + NextCap<F, PoseidonColumn>,
+{
+    let s0 = env.read_column(PoseidonColumn::State(0));
+    let s1 = env.read_column(PoseidonColumn::State(1));
+    let s2 = env.read_column(PoseidonColumn::State(2));
+    let state = [s0, s1, s2];
+    let a0 = env.read_column(PoseidonColumn::Absorb(0));
+    let a1 = env.read_column(PoseidonColumn::Absorb(1));
+    let absorb = [a0, a1];
+    let check_out = env.read_column(PoseidonColumn::Check);
+    let res = round(env, state, absorb, check_out);
+    for i in 0..=2 {
+        let next = env.read_next(PoseidonColumn::State(i));
```

Comment:
I don't understand the semantics of this. The idea behind `constrain_X` is that it can be used as a real assertion in the witness-building code. 
* When you execute `constraint_round` in the `ConstraintBuilder` mode, you'll get the constraint, alright.
* But when you execute it in the `WitnessBuilder` mode, this will fail, because ... current `WitnessBuilder` does not allow you to write into the future cells. 

I suggest this mechanic is added within this PR. You'll just need to:
* Store `cur_row` explicitly in the `WitnessBuilder`, so that `cur_row = witness.len() - 2` (e.g. witness is two rows, but current row is 0, and row 1 is "next").
* Extend capabilities of read/write column to read/write next column. 
* Add `write_column_next` to `compute_round`.